src/main/kotlin/carracing/game/controller/CarRacingController.kt

```diff
@@ -0,0 +1,63 @@
+package carracing.game.controller
+
+import carracing.game.domain.CarRacingModel
+import carracing.game.view.CarRacingView
+import carracing.game.view.ErrorView
+import carracing.game.view.InputView
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.DelicateCoroutinesApi
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+
+@OptIn(DelicateCoroutinesApi::class)
+class CarRacingController(
+    private val inputView: InputView = InputView(),
+    private val errorView: ErrorView = ErrorView(),
+    private val carRacingView: CarRacingView = CarRacingView(),
+    private val model: CarRacingModel = CarRacingModel(),
+    private val scope: CoroutineScope = GlobalScope,
+) {
+    fun startGame() {
+        getGameParameters()
+        startRacing()
+    }
+
+    private fun getGameParameters() {
+        getCarsParameter()
+        getRoundsParameter()
+    }
+
+    private fun getCarsParameter() {
+        while (true) {
+            try {
+                model.assignCarsAmount(inputView.inquireCarsAmount())
+                return
+            } catch (e: IllegalArgumentException) {
+                e.message?.let { errorView.printError(it) }
+            }
+        }
+    }
+
+    private fun getRoundsParameter() {
+        while (true) {
+            try {
+                model.assignRoundsAmount(inputView.inquireRoundsAmount())
+                return
+            } catch (e: IllegalArgumentException) {
+                e.message?.let { errorView.printError(it) }
+            }
+        }
+    }
+
+    private fun startRacing() =
+        scope.launch {
+            try {
+                model.getRaceFlow().collect {
+                    carRacingView.printCurrentRaceState(it)
+                }
+            } catch (ex: IllegalArgumentException) {
+                ex.message?.let { errorView.printError(it) }
+                startGame()
+            }
+        }
```

Comment:
If you used this to practice coroutines, that’s pretty cool! However, this course focuses more on clean code, object-oriented design, and testing rather than coroutines. So I recommend using regular loops instead. Also, the current code doesn't seem to print the game output correctly.